utils/retrieve-state/src/lib.rs

```diff
@@ -289,50 +337,369 @@ pub fn put_block_with_deploys(
+        trie_key: Digest,
+        error: bytesrepr::Error,
+    },
+}
+
+pub enum PeerMsg {
+    /// Download and save the trie under this
+    GetTrie(Digest),
+    TrieDownloaded {
+        trie: Trie<Key, StoredValue>,
+        peer: SocketAddr,
+        elapsed: Duration,
+    },
+    NoBytes {
+        trie_key: Digest,
+        peer: SocketAddr,
+    },
+}
+
 /// Download the trie from a node to the provided lmdb path.
 pub async fn download_trie(
-    client: &mut Client,
-    url: &str,
-    engine_state: &EngineState<LmdbGlobalState>,
+    client: &Client,
+    peers: &[SocketAddr],
+    engine_state: Arc<EngineState<LmdbGlobalState>>,
     state_root_hash: Digest,
-    progress_fn: impl Fn() + Send + Sync,
+    concurrent_requests_per_peer: usize,
 ) -> Result<usize, anyhow::Error> {
-    let mut outstanding_tries = vec![state_root_hash];
+    let (base_send, mut base_recv) = futures_channel::mpsc::unbounded();
+    let mut peer_futures = Vec::new();
+    for peer in peers {
+        let base_send = base_send.clone();
+        let peer_future = async move {
+            let maybe_peer =
+                match maybe_construct_peer(base_send.clone(), *peer, client, state_root_hash).await
+                {
+                    Ok(peer) => Some(peer),
+                    Err(_err) => {
+                        // just skip those that can't be communicated with
+                        // println!("error constructing peer with address {:?} {:?}", peer,
+                        // err);
+                        None
+                    }
+                };
+            maybe_peer
+        };
+        peer_futures.push(peer_future);
+    }
+    let mut peer_map = futures::future::join_all(peer_futures)
+        .await
+        .into_iter()
+        .flatten()
+        .map(|peer: Peer| (peer.address, peer))
+        .collect::<HashMap<SocketAddr, Peer>>();
 
-    let mut start = Instant::now();
-    let mut tries_downloaded = 0;
-    while let Some(next_trie_key) = outstanding_tries.pop() {
-        let read_result = get_trie(
-            client,
-            url,
-            GetTrieParams {
-                trie_key: next_trie_key,
-            },
-        )
-        .await?;
-        if let Some(blob) = read_result.maybe_trie_bytes {
-            let bytes: Vec<u8> = blob.into();
-            let (trie, _): (Trie<Key, StoredValue>, _) = FromBytes::from_bytes(&bytes)?;
-            let mut missing_descendants = engine_state
-                .put_trie_and_find_missing_descendant_trie_keys(CorrelationId::new(), &trie)?;
-            outstanding_tries.append(&mut missing_descendants);
-            tries_downloaded += 1;
-            progress_fn()
-        } else {
-            return Err(anyhow::anyhow!(
-                "unable to download trie at {:?}",
-                next_trie_key
-            ));
+    println!("got {}/{} valid peers", peer_map.len(), peers.len());
+
+    let mut total_tries_count = 0;
```

Comment:
Why not also count the root right away, instead of adding `1` at the end?

Reply:
I'm using the tries downloaded to determine when the downloading has completed, whereas if I counted it right away, I would lose that resolution.

Reply:
Yes, I'm not suggesting adding the root to `tries_downloaded` right away, just to `total_tries_count`.